pinot-tools/src/main/java/com/linkedin/pinot/tools/admin/command/MoveReplicaGroup.java

```diff
@@ -0,0 +1,434 @@
+        tableName + " -zkHost " + zkHost + " -zkPath " + zkPath +
+        (exec ? " -exec" : "");
+    return retString;
+  }
+
+  @Override
+  public void cleanup() {
+
+  }
+
+  public boolean execute()
+      throws IOException, JSONException, InterruptedException {
+    validateParams();
+
+    zkChanger = new PinotZKChanger(zkHost, zkPath);
+    this.helix = zkChanger.getHelixAdmin();
+
+    if (! isExistingTable(tableName)) {
+      LOGGER.error("Table {} does not exist", tableName);
+    }
+    List<String> srcHosts = readHostsFromFile(srcHostsFile);
+    if (srcHosts.isEmpty()) {
+      LOGGER.error("Empty list of servers. Nothing to do");
+      // this is not process error but most likely usage error
+      // exiting with status 1 so that scripts can catch this
+      System.exit(1);
+    }
+
+    LOGGER.info("Source hosts: {}", srcHosts);
+    String serverTenant = getServerTenantName(tableName) + "_OFFLINE";
+    LOGGER.debug("Using server tenant: {}", serverTenant);
+    List<String> destinationServers = readDestinationServers();
+
+    verifyNoServerOverlap(srcHosts, destinationServers);
+
+    Map<String, Map<String, String>> idealStateMap = helix.getResourceIdealState(zkPath, tableName)
+        .getRecord().getMapFields();
+
+
+    PriorityQueue<SourceSegments> segmentsToMove = getSegmentsToMoveQueue(idealStateMap, srcHosts);
+    PriorityQueue<ServerInstance> destinationServerQueue = getDestinationServerQueue(idealStateMap, destinationServers);
+
+    Map<String, Map<String, String>> proposedIdealState =
+        computeNewIdealState(idealStateMap, segmentsToMove, destinationServerQueue, srcHosts);
+    printNewIdealState(proposedIdealState);
+    printDestinationServerCounts(destinationServerQueue);
+    if (! exec) {
+      LOGGER.info("Run with -exec to apply this IdealState");
+      System.exit(0);
+    }
+
+    applyIdealState(proposedIdealState);
+    zkChanger.waitForStable(tableName);
+    return true;
+  }
+
+  private void printDestinationServerCounts(PriorityQueue<ServerInstance> destinationServerQueue) {
+    System.out.println("Number of segments per server: ");
+    for (ServerInstance instance : destinationServerQueue) {
+      System.out.println(instance.server + " : " + instance.segments);
+    }
+  }
+
+  private void printNewIdealState(Map<String, Map<String, String>> idealState)
+      throws JsonProcessingException {
+    System.out.println("Proposed idealstate is: ");
+    String output = new ObjectMapper().writerWithDefaultPrettyPrinter().writeValueAsString(idealState);
+    System.out.println(output);
+  }
+
+  private void applyIdealState(final Map<String, Map<String, String>> proposedIdealState) {
+    HelixHelper.updateIdealState(zkChanger.getHelixManager(), tableName, new Function<IdealState, IdealState>() {
+      @Nullable
+      @Override
+      public IdealState apply(@Nullable IdealState input) {
+        ZNRecord record = input.getRecord();
+        record.setMapFields(proposedIdealState);
```

Comment:
the fool proof way is to iterate over the map and set the individual entries (leaving the ones you don't care about).

Each time apply() is called, you will get the current idealstate, and you can update it to your heart's content. If it is not able to update (because of another change that happened), then apply() will get called with a newer version of idealstate